jupyterlab/handlers/yjs_echo_ws.py

```diff
@@ -119,6 +146,39 @@ def check_origin(self, origin):
     def hook_send_message(self, msg):
         self.write_message(msg, binary=True)
 
+    def maybe_save_document(self):
+        if not self.room.dirty:
+            return
+        if self.saving_document is not None and not self.saving_document.done():
+            # the document is being saved, cancel that
+            self.saving_document.cancel()
+            self.saving_document = None
+        # the document might not be already fully created, try to generate its source
+        try:
+            source = self.room.source
+        except Exception:
+            source = None
+        # save the document if we could generate its source
+        if source is not None:
+            if self.room.type == "notebook":
+                source = json.dumps(source, indent=2)
+            self.saving_document = asyncio.create_task(self.save_document(source))
+
+    async def save_document(self, source):
+        # save after 1 second of inactivity to prevent too frequent saving
+        await asyncio.sleep(1)
+        path = self.room_id
+        self.log.info(f"Saving file at /{path}")
+        async with await open_file(path, "w") as f:
+            await f.write(source)
+        # the client could be gone at that time (for instance, the user closed their browser)
+        # so try sending the "content saved" to the front-end and tolerate a closed WebSocket
+        try:
+            self.write_message(bytes([ServerMessageType.CONTENT_SAVED, 1]), binary=True)
+            # self.room.dirty = False
```

Comment:
Thanks for letting me know. Looks like this is a problem with Ypy. Looking into a fix now.

Reply:
Fixed the issue in [this PR](https://github.com/y-crdt/ypy/pull/20)

Reply:
Thanks for the quick feedback John.
I compiled ypy from source and while it fixes the boolean encoding, I can see that when I set this dirty flag in the back-end to `False`, it keeps this value forever, even though the front-end is supposed to set it back to `True` when the user modifies the document.
Sorry to talk about high-level, Jupyter-specific terms, I will try to come up with a minimum reproducible example in ypy. In the meantime, if you see something obvious in ypy, please let me know.

Reply:
Forget it, I think it's because currently the back-end is "read-only", meaning that it receives all the updates from the front-ends, but anything set with ypy is not broadcast to the front-ends through the websockets.

Reply:
Got it, we'll release a new version tomorrow so you can pull it in with conda. There are significant API updates with this version that eliminates the need for transactions in all read-only operations, replacing many of these functions with Python dunder methods. Details can be found [here](https://github.com/y-crdt/ypy/pull/16)

Reply:
Thanks, yes I saw the backward incompatibilities. BTW, what will be the next version? I think I need to pin exact versions of ypy in JupyterLab, until we reach v1.0.0. Currently we pin `y-py>=0.2.3,<1`, which will break with next release.

Reply:
@davidbrochart You can go for semver: `^0.3.0` should be the next version. We will bump the package to `0.4.0` when the next breaking change happens.